File: common/scala/src/main/scala/org/aiddl/common/scala/math/linear_algebra/NormL.scala
package org.aiddl.common.scala.math.linear_algebra

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.representation._

import org.aiddl.common.scala.math.linear_algebra.Matrix
import org.aiddl.common.scala.math.linear_algebra.AiddlMatrix
import org.aiddl.common.scala.math.linear_algebra.LupDecomposition
import org.aiddl.common.scala.math.linear_algebra.LupSolver

class NormL(n: Int) extends Function {
  def apply( x: Matrix ): Num = Num(
      Math.pow(x.iterRow(0).foldLeft(0.0)((c, t) => c + Math.pow(Math.abs(t.asNum.toDouble), n)), 1.0 / n)
    )

  def apply( x: Term ): Term = Num(Math.pow(x.asCol.foldLeft(0.0)( (c, t) => c + Math.pow(Math.abs(t.asNum.toDouble), n) ), 1.0/n))
}

File: common/java/src/main/java/org/aiddl/common/java/math/linear_algebra/NormL.java
package org.aiddl.common.java.math.linear_algebra;

import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.Term;

public class NormL implements Function {
	
	int n = 2;
	
	public NormL( int n ) {
		this.n = n;
	}
	
	public Term apply( Term x ) {
		double s = 0.0;
		double n_r = (double)n;
		for ( int i = 0 ; i < x.size() ; i++ ) {
			double v = x.get(i).getDoubleValue();
			s += Math.pow(Math.abs(v), n_r);
		}
		return Term.real(Math.pow(s, 1.0/n_r));
	}
}
